### PR TITLE
Add `ControlFlowSideEffects` parameter to `createFunction` in EVMDialect

### PR DESCRIPTION
**Pull request by rodiazet**
- This is required by eof contract creation to set control flow side effect nicely .

### CODE REVIEW

File: libyul/backends/evm/EVMDialect.cpp
Comment (cameel) [inline libyul/backends/evm/EVMDialect.cpp]: Like I mentioned in https://github.com/ethereum/solidity/pull/15512#discussion_r1809364632, I'd also extract this bit from `createEVMFunction()` into a helper: https://github.com/ethereum/solidity/blob/a80a4daaa8192d9cf5f7c164a7c591260efc3d68/libyul/backends/evm/EVMDialect.cpp#L71-L84

This would let us avoid setting them by hand and reduce the chance of getting them wrong. They can already be derived from the info we have in `SemanticInformation`, and that should be the single source of truth.

Then for any builtin based on an existing instruction you could do:
```suggestion
			ControlFlowSideEffects::fromInstruction(evmasm::Instruction::DATALOADN),
```

Comment (rodiazet) [reply libyul/backends/evm/EVMDialect.cpp]: Added the helper. One comment here. It won't be always an instruction which can be used to get proper `ControlFlowSideEffects`. If we want to use this helper with `setimmutable` for legacy dialect it's not clear which instruction should be pass to the helper.

Comment (cameel) [reply libyul/backends/evm/EVMDialect.cpp]: But the helper you added does not have the logic from `createEVMFunction()` that I linked above. Is it the final version?

> If we want to use this helper with `setimmutable` for legacy dialect it's not clear which instruction should be pass to the helper.

Yes, of course. The helper is meant only for those builtins that are based on an instruction. That's the case with `auxdataloadn()`. And will be the case with `eofcreate()` and `returncontract()`. Maybe with more of them.

For builtins that do not fit this, there's still an option of specifying default side effects or even setting them by hand. It just should not be necessary for those based on instructions.

File: libyul/backends/evm/EVMDialect.cpp
Comment (cameel) [inline libyul/backends/evm/EVMDialect.cpp]: One more thing:

```suggestion
			EVMDialect::sideEffectsOfInstruction(evmasm::Instruction::DATALOADN),
```

I think the result will be the same, but it should really be coming from `SemanticInformation` whenever possible to avoid mistakes.

File: libyul/backends/evm/EVMDialect.cpp
Comment (cameel) [inline libyul/backends/evm/EVMDialect.cpp:529]: Oh, just noticed this now - looks like this PR actually uncovered a pre-existing bug. I think this should be `ControlFlowSideEffects{true, true, true}` because `verbatim` can do virtually anything.

Can you submit a fix for this in a separate PR? This will need a changelog entry and we need to think how serious this is in practice. Maybe `SideEffects::worst()` is enough to prevent optimizer doing anything wrong with it. Maybe not. It's not relevant to EOF though, so no need for it to block merging of this PR.

Comment (rodiazet) [reply libyul/backends/evm/EVMDialect.cpp:529]: https://github.com/ethereum/solidity/pull/15542

File: libyul/backends/evm/EVMDialect.cpp
Comment (rodiazet) [reply libyul/backends/evm/EVMDialect.cpp]: fixed

File: libyul/backends/evm/EVMDialect.cpp
Comment (rodiazet) [reply libyul/backends/evm/EVMDialect.cpp]: Sorry. Too tired yesterday night. Fixed now

**Review by cameel — APPROVED** [review]

File: libyul/backends/evm/EVMDialect.cpp
Comment (cameel) [inline libyul/backends/evm/EVMDialect.cpp:319]: I wanted to suggest to use `EVMDialect::sideEffectsOfInstruction(evmasm::Instruction::CODECOPY)` for `SideEffects` here too, but apparently the `movableApartFromEffects` flag here differs from what `SemanticInformation::movableApartFromEffects()` returns for this instruction.

The discrepancy Looks like a bug to me and confirms that we should be getting these values from `SemanticInformation` as a single source of truth instead of repeating them. Unless this case is actually intentional? @ekpyron?

**Comment by cameel** [comment]
I'd merge this one already, but apparently github has not noticed that [CI passed after I reran it](https://app.circleci.com/pipelines/github/ethereum/solidity/36664/workflows/88f74e61-bd34-4834-9f54-92d31bf68863). Maybe try force pushing it to retrigger the checks (after amending it without changes just to alter the commit hash and make it look like a new one).

File: libyul/backends/evm/EVMDialect.cpp
Comment (ekpyron) [reply libyul/backends/evm/EVMDialect.cpp:319]: I'm only seeing this now - it shouldn't actually make a difference in this case, but yeah, we can use the ``SemanticInformation`` version instead.

Comment (ekpyron) [reply libyul/backends/evm/EVMDialect.cpp:319]: (as in, we probably should, the fact that it doesn't make a difference is more of a happy accident than intent here)

Comment (cameel) [reply libyul/backends/evm/EVMDialect.cpp:319]: Good. Here's a PR that fixes it: #15556.

Turns out that the flag was randomly flipped in a refactor: https://github.com/ethereum/solidity/pull/9283#discussion_r452236682. Chris actually discovered that and requested a change, and it was marked as resolved, but the fix was apparently never applied before the PR was merged.

Fortunately looks like the value of this flag does not change anything for instructions that write to memory and this is the case here, so nothing was actually broken.

**Pull request merged**
